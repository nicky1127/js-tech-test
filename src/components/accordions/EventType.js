import React, { useState } from "react";
import PropTypes from "prop-types";
import { Redirect, useHistory } from "react-router-dom";
import clsx from "clsx";
import { connect } from "react-redux";
import { makeStyles } from "@material-ui/core/styles";
import {
  Accordion,
  AccordionSummary,
  AccordionDetails,
  Typography,
  Paper,
  Table,
  TableContainer,
  TableRow,
  TableCell,
  TableBody
} from "@material-ui/core";
import { DoubleArrow } from "@material-ui/icons";
import ScoreTag from "components/units/ScoreTag";
import { getEventById } from "redux/_actions";
import labels from "constants/labels";

const constants = labels.EventType;

const useStyles = makeStyles(theme => ({
  root: {
    width: "100%",
    backgroundColor: "rgb(239,244,252)"
  },
  header: {
    color: "rgb(50,53,101)",
    backgroundColor: "rgb(208,219,233)"
  },
  table: {
    width: "100%"
  },
  row: {
    "&:hover": {
      cursor: "pointer"
    }
  },
  details: {
    padding: "0"
  },
  arrow: {
    transform: "rotate(90deg)"
  }
}));

export const EventType = props => {
  const classes = useStyles();
  const { group } = props;
  const [expanded, setExpanded] = useState(true);
  const [redirect, setRedirect] = useState(false);
  const [path, setPath] = useState("/");

  let history = useHistory();

  const rows = group[1].map(event => ({
    time: new Date(event.startTime).toLocaleTimeString("en-GB", {
      hour: "numeric",
      minute: "numeric"
    }),
    name: event.name,
    scores: event.scores,
    eventId: event.eventId
  }));

  const onClickRow = eventId => {
    history.push("/");
    setPath(`/events/${eventId}`);
    setTimeout(() => setRedirect(true), 200);
  };

  if (redirect) {
    return <Redirect to={path} />;
  }

  return (
    <div className={clsx("EventType", classes.root)} data-testid="eventType">
      <Accordion expanded={expanded}>
        <AccordionSummary
          className={classes.header}
          expandIcon={
            <DoubleArrow fontSize="small" className={classes.arrow} />
          }
          onClick={() => setExpanded(!expanded)}
          aria-controls="panel1a-content"
          id="panel1a-header"
        >
          <Typography className={classes.heading}>{group[0]}</Typography>
        </AccordionSummary>
        <AccordionDetails classes={{ root: classes.details }}>
          <TableContainer component={Paper}>
            <Table className={classes.table} aria-label="simple table">
              <TableBody>
                {rows.map((row, idx) => (
                  <TableRow
                    className={classes.row}
                    key={`${row.name}_${idx}`}
                    onClick={() => onClickRow(row.eventId)}
                  >
                    <TableCell component="th" scope="row">
                      {row.time}
                    </TableCell>
                    <TableCell align="left">{row.name}</TableCell>
                    <TableCell align="right">
                      <ScoreTag scores={row.scores} />
                    </TableCell>
                  </TableRow>
                ))}
              </TableBody>
            </Table>
          </TableContainer>
        </AccordionDetails>
      </Accordion>
    </div>
  );
};

EventType.defaultProps = {};

EventType.propTypes = { loading: PropTypes.bool, group: PropTypes.array };

const mapStateToProps = state => {
  const { loading } = state;
  return { loading };
};

const ConnectedEventType = connect(mapStateToProps, {
  getEventById
})(EventType);

export default ConnectedEventType;
