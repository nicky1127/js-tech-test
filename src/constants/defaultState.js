const state = {
  liveEvents: [],
  event: {},
  markets: {},
  isOddsDecimal: false,
  loading: false,
  error: {},
  selections: []
};

export default state;
