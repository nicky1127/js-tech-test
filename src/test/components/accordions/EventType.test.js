import React from "react";
import { render, cleanup } from "@testing-library/react";
import renderer from "react-test-renderer";
import EventType from "components/accordions/EventType";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import defaultState from "constants/defaultState";

beforeEach(() => {
  jest.clearAllMocks();
  cleanup();
});

const props = {group:['test',[{startTime:'2017-09-19T12:16:51.000Z'}]]};

const state = {
  ...defaultState
};
const mockStore = configureMockStore();
const store = mockStore(state);

describe("EventType", () => {
  it("Snapshot matchs", () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <EventType {...props} />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("EventType should render one [data-testid=eventType]", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <EventType {...props} />
      </Provider>
    );
    const container = getByTestId("eventType");
    expect(container).toBeInTheDocument();
  });
});
