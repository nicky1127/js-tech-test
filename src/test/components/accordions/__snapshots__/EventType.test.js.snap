// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventType Snapshot matchs 1`] = `
<div
  className="EventType makeStyles-root-1"
  data-testid="eventType"
>
  <div
    className="MuiPaper-root MuiAccordion-root Mui-expanded MuiAccordion-rounded MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      aria-controls="panel1a-content"
      aria-disabled={false}
      aria-expanded={true}
      className="MuiButtonBase-root MuiAccordionSummary-root makeStyles-header-2 Mui-expanded"
      id="panel1a-header"
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="MuiAccordionSummary-content Mui-expanded"
      >
        <p
          className="MuiTypography-root MuiTypography-body1"
        >
          test
        </p>
      </div>
      <div
        aria-disabled={false}
        aria-hidden={true}
        className="MuiButtonBase-root MuiIconButton-root MuiAccordionSummary-expandIcon Mui-expanded MuiIconButton-edgeEnd"
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role={null}
        tabIndex={null}
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root makeStyles-arrow-6 MuiSvgIcon-fontSizeSmall"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.5 5H11l5 7-5 7h4.5l5-7z"
            />
            <path
              d="M8.5 5H4l5 7-5 7h4.5l5-7z"
            />
          </svg>
        </span>
      </div>
    </div>
    <div
      className="MuiCollapse-container MuiCollapse-entered"
      style={
        Object {
          "minHeight": "0px",
        }
      }
    >
      <div
        className="MuiCollapse-wrapper"
      >
        <div
          className="MuiCollapse-wrapperInner"
        >
          <div
            aria-labelledby="panel1a-header"
            id="panel1a-content"
            role="region"
          >
            <div
              className="MuiAccordionDetails-root makeStyles-details-5"
            >
              <div
                className="MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded"
              >
                <table
                  aria-label="simple table"
                  className="MuiTable-root makeStyles-table-3"
                  role={null}
                >
                  <tbody
                    className="MuiTableBody-root"
                    role={null}
                  >
                    <tr
                      className="MuiTableRow-root makeStyles-row-4"
                      onClick={[Function]}
                      role={null}
                    >
                      <th
                        aria-sort={null}
                        className="MuiTableCell-root MuiTableCell-body"
                        role="cell"
                        scope="row"
                      >
                        1:16 PM
                      </th>
                      <td
                        aria-sort={null}
                        className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft"
                      />
                      <td
                        aria-sort={null}
                        className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
                      >
                        <div
                          className="score makeStyles-root-7"
                          data-testid="scoreTag"
                        >
                          undefined-undefined
                        </div>
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
